flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/ChangelogStateBackendHandle.java

```diff
@@ -237,6 +237,23 @@ public FSDataInputStream openInputStream() throws IOException {
             public Optional<byte[]> asBytesIfInMemory() {
                 throw new UnsupportedOperationException("Should not call here.");
             }
+
+            @Override
+            public boolean equals(Object o) {
+                if (this == o) {
+                    return true;
+                }
+                if (o == null || getClass() != o.getClass()) {
+                    return false;
+                }
+                StreamStateHandleWrapper that = (StreamStateHandleWrapper) o;
+                return Objects.equals(keyedStateHandle, that.keyedStateHandle);
```

Comment:
There is one more problem here - `PlaceholderStreamStateHandle`.

If there were two or more materializations, but the internal backend used a previously uploaded SST, then `equals` will return false. The wrapper will receive a discard call again.

I [used](https://github.com/apache/flink/pull/18556/commits/2403a8446da47c558d1d0c3283d056ac6f7b1988#diff-9ec20ab5bc5e16e0c04167e1b546902cdb49635ac5d15f4172fb7f36a98c77bdR250) `getStateHandleId` for comparison, and it solved the problem.

Reply:
The `PlaceholderStreamStateHandle` itself is not the materialized keyed state handle, but part of it. Actually, I did not understand in what case will we have two or more materializations, in the case of restoring?

Or in other words, what test will fail if do not compare the state handle id?

Reply:
> The PlaceholderStreamStateHandle itself is not the materialized keyed state handle, but part of it.

Yes, but because of wrapping, `SharedStateRegistry` will call `IncrementalRemoteKeyedStateHandle.equals`, which in turn will compare File state handle with a placeholder. Currently, `SharedStateRegistry` has a separate branch for placeholders.

> Actually, I did not understand in what case will we have two or more materializations, in the case of restoring?

Yes.

> Or in other words, what test will fail if do not compare the state handle id?

IIRC, `EventTimeWindowCheckpointingITCase` was failing with materialization interval of `100ms`.

Reply:
Thanks for your explanation. The deserialized `PlaceHolderStateHandle` is not equals to previous registered one.